### PR TITLE
[WIP] Add VITS-2

### PR DESCRIPTION
**Pull request by ezerhouni**
Add VITS-2 Recipe for LJSpeech.
For the moment, I am facing a bug tracked here :
https://github.com/k2-fsa/icefall/issues/1508

I will add the License etc later on, once the PR work

### CODE REVIEW

**Comment by csukuangfj** [comment]
Does `torch.min(inputs, dim=None)` work for you?

**Comment by ezerhouni** [comment]
For now, I am getting
```
  File "/vits2/egs/ljspeech/TTS/vits2/transform.py", line 118, in rational_quadratic_spline
    if torch.min(inputs, dim=None) < left or torch.max(inputs) > right:
RuntimeError: Please look up dimensions by name, got: name = None.
```

Debugging as we speak

**Comment by csukuangfj** [comment]
What is the output of
```
print(type(inputs))
```

**Comment by ezerhouni** [comment]
> What is the output of
> 
> ```
> print(type(inputs))
> ```

```<class 'torch.Tensor'>```

**Comment by csukuangfj** [comment]
As a last attempt, could you replace 
```
torch.min(inputs, dim=None)
```
with
```
torch.min(inputs.reshape(-1), dim=0)
```

---

I just searched for the new error, and found
https://github.com/pytorch/pytorch/issues/70925

Not sure if `inputs` is a named tensor.

**Comment by ezerhouni** [comment]
@csukuangfj Thanks for your help, I have replaced by:

```
    if torch.min(inputs.reshape(-1), dim=0)[0] < left or torch.max(inputs.reshape(-1), dim=0)[0] > right:
        raise ValueError("Input to a transform is not within its domain")
```

and it seems to run. It takes ~ 24h on my machine to train a model for 1000 epochs. I will let you know once the training is done

File: egs/ljspeech/TTS/vits2/residual_coupling.py [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+            g (Optional[Tensor]): Global conditioning tensor (B, global_channels, 1).
+            inverse (bool): Whether to inverse the flow.
+        Returns:
+            Tensor: Output tensor (B, in_channels, T).
+            Tensor: Log-determinant tensor for NLL (B,) if not inverse.
+        """
+        xa, xb = x.split(x.size(1) // 2, dim=1)
+
+        x_trans_mask = make_pad_mask(torch.sum(x_mask, dim=[1, 2]).type(torch.int64))
+        xa_ = self.pre_transformer(
```
Comment (ezerhouni) [inline egs/ljspeech/TTS/vits2/residual_coupling.py:363]: @csukuangfj I will try to debug more, but I think the issue might comes from there

**Comment by ezerhouni** [comment]
@csukuangfj I tried with the fix that you proposed and with the new tokenization, both throw the same error.

**Comment by csukuangfj** [comment]
Could you check that `inputs` is not empty in
```
    if torch.min(inputs.reshape(-1), dim=0)[0] < left or torch.max(inputs.reshape(-1), dim=0)[0] > right:
```
?

If it is empty, could you find why it is empty?

**Comment by ezerhouni** [comment]
@csukuangfj The training works fine. The quality is not optimal for the moment, so I am playing a bit with the parameters and will let you know

**Comment by ezerhouni** [comment]
@csukuangfj I am still having issues to train a model. I will close this PR and re-open one once I have a more stable branch.

**Pull request closed without merging**